src/EditorFeatures/Core/Tagging/AsynchronousViewportTaggerProvider.cs

```diff
@@ -0,0 +1,124 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Collections;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Tagging;
+using Microsoft.CodeAnalysis.Workspaces;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Tagging;
+
+namespace Microsoft.CodeAnalysis.Editor.Tagging;
+
+/// <summary>
+/// Base type for async taggers that perform their work based on what code is current visible in the user's
+/// viewport.  These taggers will compute tags for what is visible at their specified <see
+/// cref="AbstractAsynchronousTaggerProvider{TTag}.EventChangeDelay"/>, but will also compute tags for the regions
+/// above and below that visible section at a delay of <see cref="DelayTimeSpan.NonFocus"/>.
+/// </summary>
+internal abstract partial class AsynchronousViewportTaggerProvider<TTag> : IViewTaggerProvider
+    where TTag : ITag
+{
+    private enum ViewPortToTag
+    {
+        InView,
+        AboveAndBelow,
+    }
+
+    /// <summary>
+    /// An amount of lines above/below the viewport that we will always tag, just to ensure that scrolling up/down a few
+    /// lines shows immediate/accurate tags.
+    /// </summary>
+    private const int s_standardLineCountAroundViewportToTag = 10;
+
+    private readonly int _extraLinesAroundViewportToTag;
+    private readonly ImmutableArray<SingleViewportTaggerProvider> _viewportTaggerProviders;
+
+    protected readonly IThreadingContext ThreadingContext;
+    protected readonly IGlobalOptionService GlobalOptions;
+    protected readonly IAsynchronousOperationListener AsyncListener;
+
+    protected AsynchronousViewportTaggerProvider(
+        IThreadingContext threadingContext,
+        IGlobalOptionService globalOptions,
+        ITextBufferVisibilityTracker? visibilityTracker,
+        IAsynchronousOperationListener asyncListener,
+        int extraLinesAroundViewportToTag = 100)
```

Comment:
>100 [](http://example.com/codeflow?start=44&length=3)

I feel like if this isn't specified that it should default to some multiple of the number of lines visible in the viewport (maybe 2). Essentially, this would protect against a couple page up/downs. 100 really isn't a guarantee that it would even cover a single page up/down on a large vertical monitor.

Reply:
i'm ok with that work being done.  to me though, this is already what we're doing today, just made better.  I'm totally ok with later work continuing down this path and improving that scenario.  However, as this is no worse than today, and should strictly be better, i consider the above something to push out to later.